### PR TITLE
Added in an authorization request claims.

### PR DESCRIPTION
**Pull request by derekcollison**
These are for external authentication request from a NATS server.

Signed-off-by: Derek Collison <derek@nats.io>

### CODE REVIEW

**Review by aricart — APPROVED** [review]
LGTM

File: v2/authorization_claims.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+// ClientInformation is information about a client that is trying to authorize.
+type ClientInformation struct {
+	Host    string  `json:"host,omitempty"`
+	ID      uint64  `json:"id,omitempty"`
+	User    string  `json:"user,omitempty"`
+	Name    string  `json:"name,omitempty"`
+	Tags    TagList `json:"tags,omitempty"`
+	NameTag string  `json:"name_tag,omitempty"`
+	Kind    string  `json:"kind,omitempty"`
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply v2/authorization_claims.go:39]: Kind is internal to NATS connections, e.g. Client, Leaf, Route. First two are applicable here and match what is in connz (https://demo.nats.io:8222/connz)

Type is NATS, MQTT, WSS, etc..

File: v2/authorization_claims.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	Host string `json:"host"`
+	ID   string `json:"id"`
+}
+
+// ClientInformation is information about a client that is trying to authorize.
+type ClientInformation struct {
+	Host    string  `json:"host,omitempty"`
+	ID      uint64  `json:"id,omitempty"`
+	User    string  `json:"user,omitempty"`
+	Name    string  `json:"name,omitempty"`
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply v2/authorization_claims.go:36]: Yes what is passed on the CONNECT proto.

File: v2/authorization_claims.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	NameTag string  `json:"name_tag,omitempty"`
+	Kind    string  `json:"kind,omitempty"`
+	Type    string  `json:"type,omitempty"`
+	MQTT    string  `json:"mqtt_id,omitempty"`
+	Nonce   string  `json:"nonce,omitempty"`
+}
+
+// ClientOptions represents options that were set in the CONNECT protocol from the client
+// during authorization.
+type ClientOptions struct {
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply v2/authorization_claims.go:47]: Could be in this case, its capture from ClientOpts in the server which is a superset of all options that can be passed in a CONNECT proto.

File: v2/authorization_claims.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aricart) [reply v2/authorization_claims.go:39]: We should possibly have a set of constants here - we may already have it in the lib - then tooling can deal with it

File: v2/authorization_claims.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	Certs          StringList   `json:"certs,omitempty"`
+	VerifiedChains []StringList `json:"verified_chains,omitempty"`
+}
+
+// AuthorizationRequest represents all the information we know about the client that
+// will be sent to an external authorization service.
+type AuthorizationRequest struct {
+	Server      ServerID          `json:"server_id"`
+	Information ClientInformation `json:"client_info"`
+	Options     ClientOptions     `json:"client_opts"`
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply v2/authorization_claims.go:76]: Includes other things like language and client lib version and protocol as well.

File: v2/authorization_claims.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aricart) [reply v2/authorization_claims.go:39]: (looking for them)

File: v2/authorization_claims.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (derekcollison) [reply v2/authorization_claims.go:76]: Maybe ConnectOptions?

File: v2/claims.go [hunk 5]
```diff
@@ -53,6 +55,8 @@ func IsGenericClaimType(s string) bool {
 		fallthrough
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply v2/claims.go:55]: ok so it should be returning true then for AuthorizationRequestClaim and not false. I will fix.

[the remark this answers is not in the document]
Comment (aricart) [reply v2/claims.go:55]: that was not your bug - I was looking that we had a fallthrough and really it should be exactly what it is - if it doesn't know what it is then it should do the true - the "issue" is from before. Feel free to leave as it is

[the remark this answers is not in the document]
Comment (aricart) [reply v2/claims.go:55]: (my bad on that one) :P

**Review by aricart — APPROVED** [review]
Like ConnectOptions way better! LGTM

**Pull request merged**